internal/procedures/server/start.go

```diff
@@ -47,15 +47,49 @@ func Start(initiator common.ServerInitiator) (err error) {
 		go startCourse(course)
 	}
 
-	// Cleanup any temp dirs.
-	defer util.RemoveRecordedTempDirs()
+	return nil
+}
 
-	err = server.RunServer(initiator)
-	if err != nil {
-		return fmt.Errorf("Error during server startup sequence: '%w'.", err)
+func CleanupAndStopServer() (err error) {
```

Comment:
These is a race condition here.
Pretty much on anything that touches serverInstance/serverInstance.apiServer.

Make a normal lock and use it normally on cleanup (lock the full function).
But, you will have to be more specific for setup
(mainly because you assign serverInstance.apiServer and do setup in different functions).
I recommend making another function that assigns serverInstance.apiServer, takes skipSetup, and calls setup();
then you can lock that entire function instead of locking just chunks of code in RunAndBlock() (which you would have to do now).